# Phase 1: Data understanding

In [1]:
import pandas as pd

churn = pd.read_csv("../data/raw/customer_churn.csv")
customers = pd.read_csv("../data/raw/customers.csv")
engagement = pd.read_csv("../data/raw/customer_engagement_metrics.csv")
rfm = pd.read_csv("../data/raw/customer_rfm.csv")
orders = pd.read_csv("../data/raw/orders.csv")
billing = pd.read_csv("../data/raw/subscription_billing.csv")
tickets = pd.read_csv("../data/raw/support_tickets.csv")
campaign = pd.read_csv("../data/raw/campaign_responses.csv")

| File           | Purpose           |
| -------------- | ----------------- |
| customer_churn | Target dataset    |
| customers      | Customer details  |
| engagement     | User activity     |
| RFM            | Customer value    |
| orders         | Purchase history  |
| billing        | Subscription data |
| tickets        | Support problems  |
| campaign       | Offer history     |


Analysing Churn Data set

In [2]:
churn.head()

,customer_id,tenure_months,contract_type,monthly_spend,days_since_last_order,support_tickets_30d,total_spend,order_count,avg_order_value,payment_method,...,nps_score,satisfaction_score,churned,churn_reason,predicted_churn_prob,retention_offer_sent,offer_accepted,last_contact_date,account_manager,updated_at
0,C000001,88,biennial,369.85,284,10,29988.0,57,999.60,credit_card,...,5,3,1,service,0.660,none,0,2026-03-25,Gabrielle Davis,2025-10-19
1,C000002,76,biennial,63.69,1,1,3745.8,106,124.86,paypal,...,8,4,0,none,0.342,sms,1,2025-10-16,Ryan Munoz,2026-03-10
2,C000003,94,month-to-month,295.29,11,0,23597.1,90,786.57,ach,...,7,3,0,none,0.347,none,1,2025-10-19,Monica Herrera,2026-04-05
3,C000004,84,annual,316.97,338,13,25455.3,15,848.51,invoice,...,0,3,1,competitor,0.815,call,0,2026-03-10,Jamie Arnold,2025-09-21
4,C000005,37,annual,186.06,35,3,14235.0,73,474.50,ach,...,7,4,0,none,0.431,email,0,2026-04-05,Lisa Hensley,2025-07-06


In [3]:
churn.columns

Index(['customer_id', 'tenure_months', 'contract_type', 'monthly_spend',
       'days_since_last_order', 'support_tickets_30d', 'total_spend',
       'order_count', 'avg_order_value', 'payment_method', 'autopay',
       'num_products_owned', 'complaint_count', 'nps_score',
       'satisfaction_score', 'churned', 'churn_reason', 'predicted_churn_prob',
       'retention_offer_sent', 'offer_accepted', 'last_contact_date',
       'account_manager', 'updated_at'],
      dtype='object')

| Column                | Meaning                                           |
| --------------------- | ------------------------------------------------- |
| customer_id           | Unique customer ID                                |
| tenure_months         | Number of months customer stayed with the company |
| contract_type         | Type of customer contract or subscription         |
| monthly_spend         | Monthly amount spent by the customer              |
| days_since_last_order | Number of days since the last purchase            |
| support_tickets_30d   | Support tickets(complaints) raised in the last 30 days         |
| total_spend           | Total amount spent by the customer                |
| order_count           | Total number of orders placed                     |
| avg_order_value       | Average value of each order                       |
| payment_method        | Payment method used by the customer               |
| autopay               | Whether automatic payment is enabled              |
| num_products_owned    | Number of products/services owned                 |
| complaint_count       | Total number of complaints made                   |
| nps_score             | Net Promoter Score (customer loyalty score(1-10))       |
| satisfaction_score    | Customer satisfaction rating                      |
| churned               | Whether the customer left the company or not      |
| churn_reason          | Reason for customer leaving                       |
| predicted_churn_prob  | Predicted probability of churn                    |
| retention_offer_sent  | Whether a retention offer was sent                |
| offer_accepted        | Whether the customer accepted the offer           |
| last_contact_date     | Date of last interaction with the customer        |
| account_manager       | Manager assigned to the customer                  |
| updated_at            | Date and time when the record was last updated    |


In [4]:
churn.shape

(181175, 23)

In [5]:
churn.dtypes

customer_id               object
tenure_months              int64
contract_type             object
monthly_spend            float64
days_since_last_order      int64
support_tickets_30d        int64
total_spend              float64
order_count                int64
avg_order_value          float64
payment_method            object
autopay                    int64
num_products_owned         int64
complaint_count            int64
nps_score                  int64
satisfaction_score         int64
churned                    int64
churn_reason              object
predicted_churn_prob     float64
retention_offer_sent      object
offer_accepted             int64
last_contact_date         object
account_manager           object
updated_at                object
dtype: object

In [6]:
churn.nunique()

customer_id              150000
tenure_months                96
contract_type                 3
monthly_spend             14794
days_since_last_order       327
support_tickets_30d          19
total_spend               17804
order_count                 121
avg_order_value           17804
payment_method                4
autopay                       2
num_products_owned            8
complaint_count               7
nps_score                    11
satisfaction_score            5
churned                       2
churn_reason                  5
predicted_churn_prob        862
retention_offer_sent          4
offer_accepted                2
last_contact_date           366
account_manager           17858
updated_at                  366
dtype: int64

In [7]:
churn.isna().sum().sum()

np.int64(0)

Analysing Customer Data set

In [8]:
customers.head()

,customer_id,email,name,first_name,last_name,country,state,city,street_address,postal_code,...,loyalty_points,preferred_channel,preferred_language,phone,account_status,referral_source,email_opt_in,sms_opt_in,last_login_date,created_at
0,C000001,paulwilliams@example.org,Allison Hill,Matthew,Hernandez,BE,MN,Phamfort,945 Knox Cliff,15515,...,3786,web,de,(555)717-0570,inactive,referral,1,1,2025-08-20,2024-10-14
1,C000002,keithmelton@example.com,Noah Rhodes,Spencer,Cooper,MW,ME,Port Shawn,14996 Beth Fork Suite 566,81211,...,15809,app,es,+1-864-664-5361,active,partner,0,1,2026-05-17,2025-06-18
2,C000003,bkim@example.org,Angie Henderson,John,Barnes,DK,GA,Danamouth,551 David Bypass Suite 955,45619,...,11813,email,es,001-847-316-3787x1851,active,paid_ads,1,0,2025-10-12,2023-01-20
3,C000004,gkey@example.org,Daniel Wagner,Jonathan,Delgado,KG,PR,Johnnyland,241 Ariel Mountain,99943,...,27913,email,de,(909)947-5485x345,active,partner,0,0,2026-02-14,2026-04-23
4,C000005,piercejames@example.net,Cristian Santos,Patrick,Munoz,TO,ID,Lake Barbarabury,4254 John Route Suite 579,38850,...,7472,app,fr,(350)389-7983x71289,suspended,partner,0,0,2026-02-25,2024-03-25


In [9]:
customers.shape

(93551, 28)

In [10]:
customers.columns

Index(['customer_id', 'email', 'name', 'first_name', 'last_name', 'country',
       'state', 'city', 'street_address', 'postal_code', 'signup_date', 'age',
       'gender', 'occupation', 'marital_status', 'household_size', 'segment',
       'lifetime_value', 'loyalty_points', 'preferred_channel',
       'preferred_language', 'phone', 'account_status', 'referral_source',
       'email_opt_in', 'sms_opt_in', 'last_login_date', 'created_at'],
      dtype='object')



| Column             | Meaning                                                                          |
| ------------------ | -------------------------------------------------------------------------------- |
| marital_status     | Whether the customer is single, married, divorced, etc.                          |
| household_size     | Number of people living in the customer's family/home                            |
| segment            | Customer category such as Premium, Regular, or VIP                               |
| lifetime_value     | Total value or revenue generated by the customer over time                       |
| loyalty_points     | Reward points earned by the customer                                             |
| preferred_channel  | Customer's preferred way of communication (Email, SMS, Phone)                    |
| preferred_language | Language preferred by the customer                                               |
| phone              | Customer's phone number                                                          |
| account_status     | Current account status such as Active or Inactive                                |
| referral_source    | How the customer came to the company (Advertisement, Friend, Social Media, etc.) |
| email_opt_in       | Whether the customer agreed to receive emails                                    |
| sms_opt_in         | Whether the customer agreed to receive SMS messages                              |
| last_login_date    | The last date the customer logged into the system                                |
| created_at         | Date and time when the customer account was created    

In [11]:
customers.nunique()

customer_id           93551
email                 19215
name                  17858
first_name              685
last_name              1000
country                 194
state                    59
city                  13326
street_address        20000
postal_code           18092
signup_date            1827
age                      58
gender                    3
occupation              639
marital_status            4
household_size            6
segment                   4
lifetime_value        17804
loyalty_points        42363
preferred_channel         4
preferred_language        4
phone                 20000
account_status            3
referral_source           5
email_opt_in              2
sms_opt_in                2
last_login_date         366
created_at             1827
dtype: int64

In [12]:
customers.dtypes

customer_id            object
email                  object
name                   object
first_name             object
last_name              object
country                object
state                  object
city                   object
street_address         object
postal_code             int64
signup_date            object
age                     int64
gender                 object
occupation             object
marital_status         object
household_size          int64
segment                object
lifetime_value        float64
loyalty_points          int64
preferred_channel      object
preferred_language     object
phone                  object
account_status         object
referral_source        object
email_opt_in            int64
sms_opt_in              int64
last_login_date        object
created_at             object
dtype: object

In [13]:
customers.isna().sum()

customer_id             0
email                   0
name                    0
first_name              0
last_name               0
country               469
state                   0
city                    0
street_address          0
postal_code             0
signup_date             0
age                     0
gender                  0
occupation              0
marital_status          0
household_size          0
segment                 0
lifetime_value          0
loyalty_points          0
preferred_channel       0
preferred_language      0
phone                   0
account_status          0
referral_source         0
email_opt_in            0
sms_opt_in              0
last_login_date         0
created_at              0
dtype: int64

nulls present in country 469

Analysing Engagement Data set

In [14]:
engagement.head()


,metric_id,customer_id,last_login_date,login_frequency_30d,pages_per_session,avg_session_duration_min,mobile_app_usage_pct,email_open_rate,email_click_rate,push_notification_opt_in,...,feature_adoption_score,community_forum_posts,referrals_made,support_escalations_90d,product_tours_completed,wishlist_items,cart_abandonment_rate,promo_code_usage_90d,engagement_tier,measured_date
0,ENG00000001,C000001,2025-08-20,6,14.5,39.1,82.8,38.0,18.3,0,...,36.0,6,1,4,3,14,0.81,9,medium,2025-08-20
1,ENG00000002,C000002,2026-05-17,38,9.1,35.1,55.8,62.3,14.9,1,...,21.7,16,2,0,3,28,0.17,8,low,2026-05-17
2,ENG00000003,C000003,2025-10-12,54,1.8,21.6,41.1,52.4,16.8,1,...,52.8,6,6,0,10,18,0.30,2,dormant,2025-10-12
3,ENG00000004,C000004,2026-02-14,34,12.1,12.5,57.2,48.9,17.8,1,...,7.5,15,0,0,3,22,0.14,8,low,2026-02-14
4,ENG00000005,C000005,2026-02-25,29,14.7,38.2,50.8,26.3,12.4,1,...,30.1,11,3,0,4,28,0.09,7,high,2026-02-25


In [15]:
engagement.shape

(219237, 22)

In [16]:
engagement.columns

Index(['metric_id', 'customer_id', 'last_login_date', 'login_frequency_30d',
       'pages_per_session', 'avg_session_duration_min', 'mobile_app_usage_pct',
       'email_open_rate', 'email_click_rate', 'push_notification_opt_in',
       'nps_score', 'csat_score', 'feature_adoption_score',
       'community_forum_posts', 'referrals_made', 'support_escalations_90d',
       'product_tours_completed', 'wishlist_items', 'cart_abandonment_rate',
       'promo_code_usage_90d', 'engagement_tier', 'measured_date'],
      dtype='object')

| Column                   | Meaning                                                |
| ------------------------ | ------------------------------------------------------ |
| metric_id                | Unique engagement record ID                            |
| customer_id              | Unique customer ID                                     |
| last_login_date          | Customer's last login date                             |
| login_frequency_30d      | Number of logins in the last 30 days                   |
| pages_per_session        | Average pages viewed in one session                    |
| avg_session_duration_min | Average session duration in minutes                    |
| mobile_app_usage_pct     | Percentage of mobile app usage                         |
| email_open_rate          | Percentage of emails opened                            |
| email_click_rate         | Percentage of email links clicked                      |
| push_notification_opt_in | Whether push notifications are enabled                 |
| nps_score                | Customer loyalty score                                 |
| csat_score               | Customer satisfaction score                            |
| feature_adoption_score   | Level of product feature usage                         |
| community_forum_posts    | Number of posts in community forums                    |
| referrals_made           | Number of customers referred                           |
| support_escalations_90d  | Number of support issues escalated in the last 90 days |
| product_tours_completed  | Number of product tutorials completed                  |
| wishlist_items           | Number of items added to wishlist                      |
| cart_abandonment_rate    | Percentage of abandoned carts                          |
| promo_code_usage_90d     | Number of promo codes used in the last 90 days         |
| engagement_tier          | Customer engagement level (Low, Medium, High)          |
| measured_date            | Date when engagement metrics were recorded             |


In [17]:
engagement.dtypes

metric_id                    object
customer_id                  object
last_login_date              object
login_frequency_30d           int64
pages_per_session           float64
avg_session_duration_min    float64
mobile_app_usage_pct        float64
email_open_rate             float64
email_click_rate            float64
push_notification_opt_in      int64
nps_score                     int64
csat_score                    int64
feature_adoption_score      float64
community_forum_posts         int64
referrals_made                int64
support_escalations_90d       int64
product_tours_completed       int64
wishlist_items                int64
cart_abandonment_rate       float64
promo_code_usage_90d          int64
engagement_tier              object
measured_date                object
dtype: object

In [18]:
engagement.nunique()

metric_id                   219237
customer_id                 150000
last_login_date                366
login_frequency_30d             61
pages_per_session              141
avg_session_duration_min       431
mobile_app_usage_pct           851
email_open_rate                701
email_click_rate               191
push_notification_opt_in         2
nps_score                       11
csat_score                       5
feature_adoption_score        1001
community_forum_posts           21
referrals_made                   9
support_escalations_90d          6
product_tours_completed         11
wishlist_items                  31
cart_abandonment_rate           86
promo_code_usage_90d            16
engagement_tier                  4
measured_date                  366
dtype: int64

In [19]:
engagement.isna().sum().sum()

np.int64(0)

Analysing RFM Data set

In [20]:
rfm.head()

,customer_id,recency_days,frequency,monetary,r_score,f_score,m_score,rfm_segment
0,C000001,172,53,28880.79,5,3,4,Lost
1,C000002,184,50,8805.44,4,4,5,New
2,C000003,253,72,13881.29,1,1,4,Loyal
3,C000004,266,6,46275.30,5,5,2,Champions
4,C000005,34,39,26255.33,3,2,3,New


In [21]:
rfm.shape

(617852, 8)

In [22]:
rfm.columns

Index(['customer_id', 'recency_days', 'frequency', 'monetary', 'r_score',
       'f_score', 'm_score', 'rfm_segment'],
      dtype='object')

| Column       | Meaning                                           |
| ------------ | ------------------------------------------------- |
| customer_id  | Unique customer ID                                |
| recency_days | Number of days since the customer's last purchase |
| frequency    | Total number of purchases made by the customer    |
| monetary     | Total amount spent by the customer                |
| r_score      | Recency score assigned to the customer            |
| f_score      | Frequency score assigned to the customer          |
| m_score      | Monetary score assigned to the customer           |
| rfm_segment  | Customer segment based on RFM scores              |


In [23]:
rfm.dtypes

customer_id      object
recency_days      int64
frequency         int64
monetary        float64
r_score           int64
f_score           int64
m_score           int64
rfm_segment      object
dtype: object

In [24]:
rfm.nunique()

customer_id     150000
recency_days       365
frequency           80
monetary        574600
r_score              5
f_score              5
m_score              5
rfm_segment          6
dtype: int64

In [25]:
rfm.isna().sum().sum()

np.int64(0)

Analysing orders Data set

In [26]:
orders.head()


,order_id,customer_id,order_date,channel,total_amount,subtotal,tax_amount,shipping_cost,discount_amount,status,...,shipping_state,shipping_city,shipping_address,billing_city,carrier,tracking_number,promo_code,is_gift,fulfillment_center,delivery_date
0,O00000001,C096901,2026-04-21 02:05:49,web,963.27,847.68,57.80,30.82,19.27,shipped,...,GA,Danamouth,551 David Bypass Suite 955,Johnnyland,UPS,TRK36F5D547-166,NaN,0,FC-EAST,2026-02-25
1,O00000002,C098477,2025-08-16 12:12:19,web,501.82,441.61,30.11,16.06,10.04,shipped,...,PR,Johnnyland,241 Ariel Mountain,Lake Barbarabury,USPS,TRK4DB6C7FF-34A,VIP,0,FC-CENTRAL,2026-03-02
2,O00000003,C038004,2025-03-22 19:46:17,web,2058.40,1811.39,123.50,65.87,41.17,returned,...,ID,Lake Barbarabury,4254 John Route Suite 579,Singletonburgh,DHL,TRK42C96037-351,SAVE10,1,FC-CENTRAL,2025-10-07
3,O00000004,C060257,2025-08-31 13:43:49,store,2173.43,1912.61,130.41,69.55,43.47,pending,...,UT,Singletonburgh,6240 Allison Shoal Suite 129,New Robert,USPS,TRK0165593A-8EA,VIP,1,FC-CENTRAL,2026-03-25
4,O00000005,C057327,2025-06-20 04:25:08,store,1374.05,1209.16,82.44,43.97,27.48,returned,...,NV,New Robert,226 Yu Fall,New Kenneth,FedEx,TRKAC47C4CC-63F,SAVE10,1,FC-CENTRAL,2025-10-16


In [27]:
orders.shape

(114387, 23)

In [28]:
orders.columns

Index(['order_id', 'customer_id', 'order_date', 'channel', 'total_amount',
       'subtotal', 'tax_amount', 'shipping_cost', 'discount_amount', 'status',
       'payment_method', 'payment_last4', 'shipping_country', 'shipping_state',
       'shipping_city', 'shipping_address', 'billing_city', 'carrier',
       'tracking_number', 'promo_code', 'is_gift', 'fulfillment_center',
       'delivery_date'],
      dtype='object')

| Column             | Meaning                                |
| ------------------ | -------------------------------------- |
| order_id           | Unique order ID                        |
| customer_id        | Unique customer ID                     |
| order_date         | Date when the order was placed         |
| channel            | Platform used to place the order       |
| total_amount       | Total order amount                     |
| subtotal           | Amount before tax and shipping         |
| tax_amount         | Tax charged on the order               |
| shipping_cost      | Shipping or delivery charge            |
| discount_amount    | Discount applied to the order          |
| status             | Current order status                   |
| payment_method     | Payment method used                    |
| payment_last4      | Last 4 digits of the payment card      |
| shipping_country   | Delivery country                       |
| shipping_state     | Delivery state                         |
| shipping_city      | Delivery city                          |
| shipping_address   | Delivery address                       |
| billing_city       | Billing city                           |
| carrier            | Delivery company or courier service    |
| tracking_number    | Shipment tracking number               |
| promo_code         | Coupon or promo code used              |
| is_gift            | Whether the order was marked as a gift |
| fulfillment_center | Warehouse that processed the order     |
| delivery_date      | Date when the order was delivered      |


In [29]:
orders.dtypes

order_id               object
customer_id            object
order_date             object
channel                object
total_amount          float64
subtotal              float64
tax_amount            float64
shipping_cost         float64
discount_amount       float64
status                 object
payment_method         object
payment_last4           int64
shipping_country       object
shipping_state         object
shipping_city          object
shipping_address       object
billing_city           object
carrier                object
tracking_number        object
promo_code             object
is_gift                 int64
fulfillment_center     object
delivery_date          object
dtype: object

In [30]:
orders.nunique()

order_id              114387
customer_id            80183
order_date             19997
channel                    4
total_amount           17804
subtotal               17804
tax_amount             10620
shipping_cost           7271
discount_amount         4872
status                     5
payment_method             4
payment_last4           8643
shipping_country         194
shipping_state            59
shipping_city          13326
shipping_address       20000
billing_city           13326
carrier                    4
tracking_number        20000
promo_code                 3
is_gift                    2
fulfillment_center         3
delivery_date            366
dtype: int64

In [31]:
orders.isna().sum()

order_id                  0
customer_id               0
order_date                0
channel                   0
total_amount              0
subtotal                  0
tax_amount                0
shipping_cost             0
discount_amount           0
status                    0
payment_method            0
payment_last4             0
shipping_country        573
shipping_state            0
shipping_city             0
shipping_address          0
billing_city              0
carrier                   0
tracking_number           0
promo_code            28616
is_gift                   0
fulfillment_center        0
delivery_date             0
dtype: int64

have many nulls in promo code

Analysing billing Data set

In [32]:
billing.head()


,billing_id,customer_id,plan_name,contract_type,monthly_recurring_revenue,annual_contract_value,billing_cycle,payment_method,autopay_enabled,paperless_billing,...,discount_pct,tenure_months,contract_end_date,payment_failures_90d,billing_address_city,tax_exempt,invoice_delivery,account_manager,created_at,updated_at
0,BIL00000001,C000001,Reverse-engineered static implementation,month-to-month,399.91,4824.00,quarterly,credit_card,0,0,...,0.8,79,2026-02-14,1,Phamfort,1,email,Noah Rhodes,2024-10-14,2025-08-20
1,BIL00000002,C000002,Extended systematic task-force,biennial,66.19,652.45,monthly,ach,0,0,...,15.3,3,2026-02-25,0,Port Shawn,0,portal,Angie Henderson,2025-06-18,2026-05-17
2,BIL00000003,C000003,Stand-alone incremental algorithm,annual,184.11,2126.55,annual,ach,1,1,...,5.3,35,2026-03-02,0,Danamouth,1,mail,Daniel Wagner,2023-01-20,2025-10-12
3,BIL00000004,C000004,Configurable non-volatile capability,month-to-month,110.28,1203.65,monthly,credit_card,1,1,...,10.4,53,2025-10-07,0,Johnnyland,0,portal,Cristian Santos,2026-04-23,2026-02-14
4,BIL00000005,C000005,Profit-focused encompassing methodology,annual,359.33,4316.80,monthly,paypal,0,0,...,16.1,47,2026-03-25,2,Lake Barbarabury,0,portal,Connie Lawrence,2024-03-25,2026-02-25


In [33]:
billing.shape

(126941, 22)

In [34]:
billing.columns

Index(['billing_id', 'customer_id', 'plan_name', 'contract_type',
       'monthly_recurring_revenue', 'annual_contract_value', 'billing_cycle',
       'payment_method', 'autopay_enabled', 'paperless_billing',
       'late_payments_12m', 'price_increase_last_year', 'discount_pct',
       'tenure_months', 'contract_end_date', 'payment_failures_90d',
       'billing_address_city', 'tax_exempt', 'invoice_delivery',
       'account_manager', 'created_at', 'updated_at'],
      dtype='object')

| Column                    | Meaning                                          |
| ------------------------- | ------------------------------------------------ |
| billing_id                | Unique billing record ID                         |
| customer_id               | Unique customer ID                               |
| plan_name                 | Subscription or service plan name                |
| contract_type             | Type of customer contract                        |
| monthly_recurring_revenue | Monthly revenue generated from the customer      |
| annual_contract_value     | Total yearly contract value                      |
| billing_cycle             | Billing period (Monthly, Quarterly, Yearly)      |
| payment_method            | Method used for payment                          |
| autopay_enabled           | Whether automatic payment is enabled             |
| paperless_billing         | Whether electronic billing is enabled            |
| late_payments_12m         | Number of late payments in the last 12 months    |
| price_increase_last_year  | Whether the price increased during the last year |
| discount_pct              | Discount percentage given to the customer        |
| tenure_months             | Number of months the customer has stayed         |
| contract_end_date         | Contract expiration date                         |
| payment_failures_90d      | Number of payment failures in the last 90 days   |
| billing_address_city      | Customer billing city                            |
| tax_exempt                | Whether the customer is exempt from tax          |
| invoice_delivery          | Method of invoice delivery                       |
| account_manager           | Manager assigned to the customer account         |
| created_at                | Record creation date and time                    |
| updated_at                | Record last updated date and time                |


In [35]:
billing.dtypes

billing_id                    object
customer_id                   object
plan_name                     object
contract_type                 object
monthly_recurring_revenue    float64
annual_contract_value        float64
billing_cycle                 object
payment_method                object
autopay_enabled                int64
paperless_billing              int64
late_payments_12m              int64
price_increase_last_year       int64
discount_pct                 float64
tenure_months                  int64
contract_end_date             object
payment_failures_90d           int64
billing_address_city          object
tax_exempt                     int64
invoice_delivery              object
account_manager               object
created_at                    object
updated_at                    object
dtype: object

In [36]:
billing.nunique()

billing_id                   126941
customer_id                  126941
plan_name                     19819
contract_type                     3
monthly_recurring_revenue     15177
annual_contract_value         17804
billing_cycle                     3
payment_method                    4
autopay_enabled                   2
paperless_billing                 2
late_payments_12m                 7
price_increase_last_year          2
discount_pct                    251
tenure_months                    84
contract_end_date               366
payment_failures_90d              4
billing_address_city          13326
tax_exempt                        2
invoice_delivery                  3
account_manager               17858
created_at                     1827
updated_at                      366
dtype: int64

In [37]:
billing.isna().sum().sum()

np.int64(0)

Analysing tickets Data set

In [38]:
tickets.head()

,ticket_id,customer_id,order_id,category,priority,status,subject,description,created_date,resolution_time_hours
0,T0000001,C113536,O00515600,account,medium,resolved,Now away land attorney many your play bank bes...,Meeting campaign interview give call tend it.,2026-03-02,65
1,T0000002,C128564,O00140908,account,medium,open,Food phone loss out option husband only talk a...,Sell represent box worry. Treatment history cu...,2025-10-07,38
2,T0000003,C080829,O00459807,billing,critical,escalated,Article national sound try citizen trouble thu...,Live majority or full model smile key. View se...,2026-03-25,70
3,T0000004,C056793,O00075875,product,high,resolved,Concern president hope radio answer policy med...,Beyond reach beyond until decade. Our federal ...,2025-10-16,18
4,T0000005,C022297,O00695391,billing,high,resolved,Instead parent far grow class own continue wha...,Apply among listen exist.,2025-10-19,48


In [39]:
tickets.shape

(98754, 10)

In [40]:
tickets.columns

Index(['ticket_id', 'customer_id', 'order_id', 'category', 'priority',
       'status', 'subject', 'description', 'created_date',
       'resolution_time_hours'],
      dtype='object')

| Column                | Meaning                                    |
| --------------------- | ------------------------------------------ |
| ticket_id             | Unique support ticket ID                   |
| customer_id           | Unique customer ID                         |
| order_id              | Order related to the support ticket        |
| category              | Type of issue or complaint                 |
| priority              | Importance level of the ticket             |
| status                | Current status of the ticket               |
| subject               | Short title of the issue                   |
| description           | Detailed explanation of the issue          |
| created_date          | Date when the ticket was created           |
| resolution_time_hours | Number of hours taken to resolve the issue |


In [41]:
tickets.dtypes

ticket_id                object
customer_id              object
order_id                 object
category                 object
priority                 object
status                   object
subject                  object
description              object
created_date             object
resolution_time_hours     int64
dtype: object

In [42]:
tickets.nunique()

ticket_id                98754
customer_id              72253
order_id                 92849
category                     5
priority                     4
status                       4
subject                  20000
description              20000
created_date               366
resolution_time_hours       72
dtype: int64

In [43]:
tickets.isna().sum().sum()

np.int64(0)

Analysing campaign Data set

In [44]:
campaign.head()

,response_id,campaign_id,customer_id,responded,converted,revenue
0,CR00000001,CMP02950,C101095,1,0,834.66
1,CR00000002,CMP00692,C132367,1,0,571.44
2,CR00000003,CMP01635,C110361,1,1,692.89
3,CR00000004,CMP02255,C079868,0,0,276.69
4,CR00000005,CMP02561,C029179,1,1,200.92


In [45]:
campaign.shape

(578395, 6)

In [46]:
campaign.columns

Index(['response_id', 'campaign_id', 'customer_id', 'responded', 'converted',
       'revenue'],
      dtype='object')

| Column      | Meaning                                                                 |
| ----------- | ----------------------------------------------------------------------- |
| response_id | Unique response record ID                                               |
| campaign_id | Unique marketing campaign ID                                            |
| customer_id | Unique customer ID                                                      |
| responded   | Whether the customer responded to the campaign                          |
| converted   | Whether the customer accepted the offer or completed the desired action |
| revenue     | Revenue generated from the campaign response                            |



In [47]:
campaign.dtypes

response_id     object
campaign_id     object
customer_id     object
responded        int64
converted        int64
revenue        float64
dtype: object

In [48]:
campaign.nunique()

response_id    578395
campaign_id      5000
customer_id    146833
responded           2
converted           2
revenue        108376
dtype: int64

In [49]:
campaign.isna().sum().sum()

np.int64(0)

| Item          | Value                 | Explanation                     |
| ------------- | --------------------- | ------------------------------- |
| Target Column | churned               | Column to be predicted          |
| Problem Type  | Binary Classification | Two possible classes            |
| 0             | Customer retained     | Customer stays with the company |
| 1             | Customer churned      | Customer leaves the company     |


| Dataset                     | Null Columns                 | Missing Values |
| --------------------------- | ---------------------------- | -------------- |
| customer_churn              | No missing values            | 0              |
| customers                   | country                      | 469            |
| customer_engagement_metrics | No missing values            | 0              |
| customer_rfm                | No missing values            | 0              |
| orders                      | shipping_country, promo_code | 573, 28616     |
| subscription_billing        | No missing values            | 0              |
| support_tickets             | No missing values            | 0              |
| campaign_responses          | No missing values            | 0              |
